Riot/Modules/Settings/DeactivateAccount/DeactivateAccountViewController.m

```diff
@@ -331,19 +304,78 @@ - (IBAction)forgetMessagesButtonAction:(UIButton*)sender
 
 - (IBAction)deactivateAccountButtonAction:(id)sender
 {
-    __weak typeof(self) weakSelf = self;
+    [self startLoading];
     
-    [self presentPasswordRequiredAlertWithSubmitHandler:^(NSString *password) {
-        
-        typeof(weakSelf) strongSelf = weakSelf;
+    MXWeakify(self);
+    [self.deactivateAccountService checkAuthenticationWithSuccess:^(enum DeactivateAccountAuthentication authentication, NSURL * _Nullable fallbackURL) {
+        MXStrongifyAndReturnIfNil(self);
         
-        if (strongSelf)
-        {
-            NSString *userId = strongSelf.mainSession.myUser.userId;
-            [strongSelf deactivateAccountWithUserId:userId andPassword:password eraseAllMessages:strongSelf.forgetMessageButton.isEnabled];
+        switch (authentication) {
+            case DeactivateAccountAuthenticationAuthenticated:
+                MXLogDebug(@"[DeactivateAccountViewController] Deactivation endpoint has already been authenticated. Continuing deactivation.")
+                [self.deactivateAccountService deactivateWithEraseAccount:self.forgetMessageButton.isEnabled];
+                break;
+            case DeactivateAccountAuthenticationRequiresPassword:
+                [self presentPasswordPrompt];
+                break;
+            case DeactivateAccountAuthenticationRequiresFallback:
+                if (fallbackURL) [self presentFallbackForURL:fallbackURL];
+                break;
         }
-        
-    } cancelHandler:nil];
+    } failure:^(NSError * _Nonnull error) {
+        MXStrongifyAndReturnIfNil(self);
+        [self handleError:error];
+    }];
+}
+
+#pragma mark - Password
+
+- (void)presentPasswordPrompt
+{
+    MXLogDebug(@"[DeactivateAccountViewController] Show password prompt.")
+    
+    MXWeakify(self);
+    [self presentPasswordRequiredAlertWithSubmitHandler:^(NSString *password) {
+        MXStrongifyAndReturnIfNil(self);
+        [self.deactivateAccountService deactivateWith:password eraseAccount:self.forgetMessageButton.isEnabled];
```

Comment:
I wonder why are we using `isEnabled` here for the `forgetMessageButton`. I feel like it should be `isSelected` instead.

Reply:
Oh gosh yeah, I completely missed that. At least we've always been deleting rather than not deleting.